# General
All ATAC and ChIP-seq scripts are numerated in order of execution, and are divided in:
    * Scripts built to be run in a SLURM queueing system (inside the folder named "01_cluster")
    * Scripts built to be run in a workstation (inside the folder named "02_workstation")

# ChIP-seq data analysis
## FILE NAMING FORMAT
[cellType]_[chip]_[date]_[extra?]._R[12]_001.fastq.gz
Only the first three sections between '_' will be used for file naming (two are also accepted)
chip can contain sections separated by '-', and only the first section will be used to count chip as a specific tag

## Scripts
### Scripts designed to run in a SLURM queueing system
    * 01_Chip_fqToBw.sh contains all the steps to go from Fastq files to filtered BAM files. Specifically, the steps to generate a summary file, Trim Fastq files, Align them, remove duplicates, remove blacklist regions and unwanted chromosomes, sort and indexing, and BigWig generation. The scripts expects three inputs:
        *  The path to the Fastq files. It is expected to contain a demux_fastq folder with the fastq files and the samplesNames.txt file (instructions in the script). Temporal analysis folders will be created in it during the run
        * The output path
        * The base name (with full path) to the reference genome. Chromosome size and blacklist files are expected to be in the same path, sharing the same base name, but ending by ".sizes" and ".blacklist.bed" respectively
    * 02b_peakAnalysis_I.sh contains the steps to go from BAM files to individual and consensus peak file and tables, respectively. Specifically, the steps to call narrow and broad peaks, annotate them, peak QC, consensus peak analysis, consensus annotation, read count in consensus peaks, CPM calculation, final table generation. The script expects one input:
        * The output path indicated in 01_ChIP_fqToBw.sh
        * Specific BAM input and peak output paths are defined in the script (allways relative to the indicated output path)
    * 02d_replicatePeakAnalysis.sh contains the steps to do differnetial enrichment analysis given that we have replicates. For that, we must include in the code the path to the non-merged bamfiles
    * Scripts containing NR are called bu the previous three scripts
### Scripts designed to run in a workstation
    * 03_getFCvals.py contains code to create a consensus peak table with the FC values between ChIP-seq experiments with the same tag but different celltype. The script expects no input, but three variables need to be modified inside:
        * "tableAll" should point to the path were the consensus peak tables generated by 02b_peakAnalysis_I are located
        * "outFC" should point the output path for the new tables
        * "deseqP" should point to the path with the DESeq differential analysis results from 02d_replicatePeakAnalysis.sh

# ATAC data analysis
## FILE NAMING FORMAT
[cellType]-[koInfo]_[batch]_[date]_[extra?]._R[12]_001.fastq.gz
Only the first three sections between '_' will be used for file naming (two are also accepted)
koInfo must contain the label of the KO gene or a label to define it as a control.

## Scripts
### Scripts designed to run in a SLURM queueing system
    * 01_ATAC_fqToBw.sh contains all the steps to go from Fastq files to filtered BAM files. Specifically, the steps to generate a summary file, Trim Fastq files, Align them, remove duplicates, remove blacklist regions and unwanted chromosomes, Tn5 adaptor removal, sort and indexing, and BigWig generation. The scripts expects three inputs:
        *  The path to the Fastq files. It is expected to contain a demux_fastq folder with the fastq files and the samplesNames.txt file (instructions in the script). Temporal analysis folders will be created in it during the run
        * The output path
        * The base name (with full path) to the reference genome. Chromosome size and blacklist files are expected to be in the same path, sharing the same base name, but ending by ".sizes" and ".blacklist.bed" respectively
    * 02b_peakAnalysis_I.sh contains the steps to go from BAM files to individual and consensus peak file and tables, respectively. Specifically, the steps to call narrow and broad peaks, annotate them, peak QC, consensus peak analysis, consensus annotation, read count in consensus peaks, CPM calculation, final table generation. The script expects one input:
        * The output path indicated in 01_ChIP_fqToBw.sh
        * Specific BAM input and peak output paths are defined in the script (allways relative to the indicated output path)
    * 02d_replicatePeakAnalysis.sh contains the steps to do differnetial enrichment analysis given that we have replicates. For that, we must include in the code the path to the non-merged bamfiles, and the posible labels (comma separated) of the controls. By default if does comparisons by batches
    * Scripts containing NR are called bu the previous three scripts
### Scripts designed to run in a workstation
    * 03_getFCvals.py contains code to create a consensus peak table with the FC values between ChIP-seq experiments with the same tag but different celltype. The script expects no input, but some variables need to be modified inside. The most important ones:
        * "tableAll" should point to the path were the consensus peak tables generated by 02b_peakAnalysis_I are located
        * "outFC" should point the output path for the new tables
        * "deseqP" should point to the path with the DESeq differential analysis results from 02d_replicatePeakAnalysis.sh
        * "controlId" contains a list of labels that can be used as controls
    * 04a_HOMERmotifs_to_TOBIASformat.sh converts HOMER motif files to a format and naming suitable for TOBIAS. The script expects no input, but some variables need to be modified inside:
        * "R" should point to the executable of R
        * "outdir" should point to the output directory
        * "motifsCheck" should point to the motifs file from HOMER
    * 04b_TOBIAS.sh runs TOBIAS. The script expects no input, but many variables need to be modified inside. More information in the script
